util/encoding/float.go

```diff
@@ -12,57 +12,50 @@
-// on sqlite4's key encoding:
-// http://sqlite.org/src4/doc/trunk/www/key_encoding.wiki
-//
 // Author: Andrew Bonventre (andybons@gmail.com)
 // Author: Peter Mattis (peter@cockroachlabs.com)
 
 package encoding
 
 import (
-	"bytes"
-	"fmt"
 	"math"
-	"strconv"
-	"unsafe"
 
 	"github.com/cockroachdb/cockroach/util"
 )
 
 // EncodeFloatAscending returns the resulting byte slice with the encoded float64
-// appended to b.
-//
-// Values are classified as large, medium, or small according to the value of
-// E. If E is 11 or more, the value is large. For E between 0 and 10, the value
-// is medium. For E less than zero, the value is small.
-//
-// Large positive values are encoded as a single byte 0x22 followed by E as a
-// varint and then M. Medium positive values are a single byte of 0x17+E
-// followed by M. Small positive values are encoded as a single byte 0x16
-// followed by the ones-complement of the varint for -E followed by M.
+// appended to b. The encoded format for a float64 value f is, for positive f, the
+// encoding of the 64 bits (in IEEE 754 format) re-interpreted as an int64 and
+// encoded using EncodeUint64Ascending. For negative f, we keep the sign bit and
+// invert all other bits, encoding this value using EncodeUint64Descending. This
+// approach was inspired by in github.com/google/orderedcode/orderedcode.go.
 //
-// Small negative values are encoded as a single byte 0x14 followed by -E as a
-// varint and then the ones-complement of M. Medium negative values are encoded
-// as a byte 0x13-E followed by the ones-complement of M. Large negative values
-// consist of the single byte 0x08 followed by the ones-complement of the
-// varint encoding of E followed by the ones-complement of M.
+// One of five single-byte prefix tags are appended to the front of the encoding.
+// These tags enforce logical ordering of keys for both ascending and descending
+// encoding directions. The tags split the encoded floats into five categories:
+// - NaN for an ascending encoding direction
+// - Negative valued floats
+// - Zero (positive and negative)
+// - Positive valued floats
+// - NaN for a descending encoding direction
+// This ordering ensures that NaNs are always sorted first in either encoding
+// direction, and that after them a logical ordering is followed.
 func EncodeFloatAscending(b []byte, f float64) []byte {
 	// Handle the simplistic cases first.
 	switch {
 	case math.IsNaN(f):
 		return append(b, floatNaN)
-	case math.IsInf(f, 1):
-		return append(b, floatInfinity)
-	case math.IsInf(f, -1):
-		return append(b, floatNegativeInfinity)
 	case f == 0:
 		return append(b, floatZero)
 	}
-	// TODO(nvanbenschoten) Switch to a base-256 mantissa.
-	e, m := floatMandE(f, b[len(b):])
-	return encodeMandE(b, f < 0, e, m)
+	i := int64(math.Float64bits(f))
+	if i < 0 {
+		i = math.MinInt64 - i
+		b = append(b, floatNeg)
+		return EncodeUint64Descending(b, uint64(-i))
```

Comment:
The subtraction, negating and inverting seems excessive here. Can't we directly call `EncodeUint64Descending(b, uint64(i))`? Which is really just a call to `EncodeUint64Ascending(b, ^uint64(i))`? I'm thinking this should look like:

```
u := math.Float64bits(f)
if u&(1<<63) != 0 {
  u = ^u
  b = append(b, floatNeg)
} else {
  b = append(b, floatPos)
}
return EncodeUint64Ascending(b, u)
```